pkg/features/kube_features.go

```diff
@@ -739,7 +740,7 @@ func init() {
 // available throughout Kubernetes binaries.
 var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	AppArmor:             {Default: true, PreRelease: featuregate.Beta},
-	DynamicKubeletConfig: {Default: true, PreRelease: featuregate.Beta},
+	DynamicKubeletConfig: {Default: false, PreRelease: featuregate.Deprecated}, // feature gate is deprecated in 1.22, remove no early than 1.23
```

Comment:
This is a breaking change. We should provide the deprecation warning for at least one release before we disable the feature gate. The deprecation policy has gotten even more complex since I last read it 😁 , but I think this is the part we want (where setting to `false` is the last thing we do, but we can set `featuregate.Deprecated` now):

> In some cases, removing pre-GA features requires considerable time. Feature gates can remain operational until their associated feature is fully removed, at which point the feature gate itself can be deprecated.

> Versioning for feature gates is different from the previously discussed components, therefore the rules for deprecation are as follows:

> Rule #8: Feature gates must be deprecated when the corresponding feature they control transitions a lifecycle stage as follows. Feature gates must function for no less than:

> Beta feature to EOL: 3 months or 1 release (whichever is longer)

> Rule #9: Deprecated feature gates must respond with a warning when used. When a feature gate is deprecated it must be documented in both in the release notes and the corresponding CLI help. Both warnings and documentation must indicate whether a feature gate is non-operational.

We should also mark the `Node.Spec.ConfigSource` field deprecated: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L4610

FWIW, we might want to consider a 2 or 3 release timeline since that field is in core/v1 (it's technically beta, and the doc comment in core/v1 still mentions a feature gate is required, but might still be best to play it safe). @liggitt WDYT?

Reply:
I read these rules as the feature gate should remain operational, but it says nothing about the defaults. So I kept it operational and for increased visibility into the change - disabled it. I understand that on the cluster upgrade, if wasn't re-enabled explicitly, it will stop working. But this is a beta feature and my assumption was that "Features can be removed at any point in the life cycle prior to GA". I'm happy to revert the change of default, I see arguments both ways.

I will add deprecation message to the fields, good catch. I was also thinking if `kubectl` should warn when these fields are used, but didn't find any precedence doing so. WDYT?

Reply:
> We should provide the deprecation warning for at least one release before we disable the feature gate.

Someone who still needs to use it for the duration of the deprecation period will turn it on explicitly in 1.22, right? Someone who doesn't will not start accidentally using it in 1.22 without realizing it is beta. I would tend to change the default to false early to increase visibility.

Reply:
> I would tend to change the default to false early to increase visibility.

I suppose that's fair, since it's most of our warning mechanisms (logs, help text, emails to kubernetes-dev) go ignored anyway...